<center><h1> Empirical Assignment 1</h1></center>

**Author(s):**
1. Belicia Rodriguez (belicia.rodriguez@emory.edu)

**Objectives**: This <ins>assignment</ins> aims at
 1. Making students familiar with *data wrangling* with the ```dplyr``` and ```tidyr``` R packages;
 2. Use *GitHub* to retrieve and submit computer code.

**Instructions**:
 1. Start by reading the [Data Wrangling with dplyr and tidyr cheatsheet](https://rstudio.com/wp-content/uploads/2015/02/data-wrangling-cheatsheet.pdf)
 2. Please write down your R code and <ins>execute</ins> it in the cell below each question.

<center><h2> Questions</h2></center>

1. Install and load the ```wooldridge```, ```dplyr```, and ```tidyr``` packages and their ```dependencies```.

In [2]:
install.packages(c('wooldridge', 'dplyr', 'tidyr'))
library(wooldridge)
library(dplyr)
library(tidyr)


  There is a binary version available but the source version is later:
      binary source needs_compilation
dplyr  0.8.3  0.8.4              TRUE

  Binaries will be installed
package 'wooldridge' successfully unpacked and MD5 sums checked
package 'dplyr' successfully unpacked and MD5 sums checked
package 'tidyr' successfully unpacked and MD5 sums checked

The downloaded binary packages are in
	C:\Users\Juan Estrada\AppData\Local\Temp\Rtmp8aJpx8\downloaded_packages


Warning message:
"package 'wooldridge' was built under R version 3.6.2"Warning message:
"package 'dplyr' was built under R version 3.6.2"
Attaching package: 'dplyr'

The following objects are masked from 'package:stats':

    filter, lag

The following objects are masked from 'package:base':

    intersect, setdiff, setequal, union

Warning message:
"package 'tidyr' was built under R version 3.6.2"

#### <font color='red'>Comments:</font> 

As I mentioned in the suggested solution, it is a good practice to use ```!require``` and the ```if``` statement to facilitate sharing code. A for loop is also a good practice when you have many packages to install.

2. Load the ```ceosal2``` data set from the ```wooldridge``` R library and create a ```tbl``` class local copy.

In [3]:
data(ceosal2, package='wooldridge')
ceo_salary <- ceosal2
head(ceo_salary, 10)

salary age college grad comten ceoten sales profits mktval lsalary  lsales  
1  1161   49  1       1     9      2      6200 966     23200  7.057037 8.732305
2   600   43  1       1    10     10       283  48      1100  6.396930 5.645447
3   379   51  1       1     9      3       169  40      1100  5.937536 5.129899
4   651   55  1       0    22     22      1100 -54      1000  6.478509 7.003066
5   497   44  1       1     8      6       351  28       387  6.208590 5.860786
6  1067   64  1       1     7      7     19000 614      3900  6.972606 9.852194
7   945   59  1       0    35     10       536  24       623  6.851185 6.284134
8  1261   63  1       1    32      8      4800 191      2100  7.139660 8.476371
9   503   47  1       1     4      4       610   7       454  6.220590 6.413459
10 1094   64  1       1    39      5      2900 230      3900  6.997596 7.972466
   lmktval   comtensq ceotensq profmarg 
1  10.051908   81       4      15.580646
2   7.003066  100     100      16.961130
3   7.003066   81       9      23.668638
4   6.907755  484     484      -4.909091
5   5.958425   64      36       7.977208
6   8.268732   49      49       3.231579
7   6.434546 1225     100       4.477612
8   7.649693 1024      64       3.979167
9   6.118097   16      16       1.147541
10  8.268732 1521      25       7.931035

#### <font color='red'>Comments:</font> 

To create a ```tbl``` class local copy you need to use the ```tbl_df``` function from the ```dplyr``` package.

3. Estimate $E[\texttt{comten}|\texttt{college}=1]$, $E[\texttt{comten}|\texttt{college}=0]$, $E[\texttt{ceoten}|\texttt{college}=1]$ and $E[\texttt{ceoten}|\texttt{college}=0]$, save these values and display them.

In [4]:
# Estimate E[comten|college=1]
comten_college1 <- mean((filter(ceo_salary, college==1))$comten)
round(comten_college1, 3)

# Estimate E[comten|college=0]
(comten_college0 <- mean((filter(ceo_salary, college==0))$comten))

# Estimate E[ceoten|college=1]
ceoten_college1 <- mean((filter(ceo_salary, college==1))$ceoten)
round(ceoten_college1,3)

# Estimate E[ceoten|college=0]
(ceoten_college0 <- mean((filter(ceo_salary, college==0))$ceoten))

[1] 22.174

[1] 33.8

[1] 7.826

[1] 12.4

4. Add two *new* variables to your local copy of the data set, ```d_comten_sq``` and ```d_ceoten_sq```, defined as the square of the difference between the corresponding ```comten``` and ```ceoten``` values and the corresponding means you calculated in (3) above, i.e., if $\texttt{comten}_i$ corresponds to the years with the company of CEO $i$ who went to college, then the corresponding value of $\texttt{d}\_\texttt{comten}\_\texttt{sq}_i$ should be $(\texttt{comten}_i-\widehat{E}[\texttt{comten}|\texttt{college}=1])^2$, where $\widehat{E}[\texttt{comten}|\texttt{college}=1]$ represents the estimate of $E[\texttt{comten}|\texttt{college}=1]$ you obtained in (3) above for example.

In [5]:
# define variables as square of difference between corresponding values and mean

# create d_comten_sq and d_ceoten_sq variables
ceo_salary$d_comten_sq <- rep(NA, nrow(ceo_salary))
ceo_salary$d_ceoten_sq <- rep(NA, nrow(ceo_salary))

# define loop through all observations
for(i in 1:nrow(ceo_salary)) {
    # get value in comten and ceoten row
    comten_i <- ceo_salary$comten[i]
    ceoten_i <- ceo_salary$ceoten[i]
    
    # subtract value and previously calculated mean
    comten_i <- comten_i - comten_college1
    ceoten_i <- ceoten_i - ceoten_college1
    
    # square the difference and add to dataset
    ceo_salary$d_comten_sq[i] <- comten_i^2
    ceo_salary$d_ceoten_sq[i] <- ceoten_i^2
}

head(ceo_salary)


salary age college grad comten ceoten sales profits mktval lsalary  lsales  
1 1161   49  1       1     9      2      6200 966     23200  7.057037 8.732305
2  600   43  1       1    10     10       283  48      1100  6.396930 5.645447
3  379   51  1       1     9      3       169  40      1100  5.937536 5.129899
4  651   55  1       0    22     22      1100 -54      1000  6.478509 7.003066
5  497   44  1       1     8      6       351  28       387  6.208590 5.860786
6 1067   64  1       1     7      7     19000 614      3900  6.972606 9.852194
  lmktval   comtensq ceotensq profmarg  d_comten_sq  d_ceoten_sq
1 10.051908  81        4      15.580646 173.56530557  33.9373986
2  7.003066 100      100      16.961130 148.21646836   4.7280963
3  7.003066  81        9      23.668638 173.56530557  23.2862358
4  6.907755 484      484      -4.909091   0.03042185 200.9141428
5  5.958425  64       36       7.977208 200.91414278   3.3327474
6  8.268732  49       49       3.231579 230.26297999   0.6815846

#### <font color='red'>Comments:</font> 

As mentioned in the suggested answer, it is important to use vectorized functions instead of loops. In this case, the function ```case_when()``` from ```dplyr``` is a good choice. Also, I do not see where you are using the values *comten_college0* and *ceoten_college0* to define the variables.

5. Print the sample median of ```d_comten_sq``` and ```d_ceoten_sq```.

In [7]:
round(median(ceo_salary$d_comten_sq),3)
round(median(ceo_salary$d_ceoten_sq),3)

[1] 117.193

[1] 23.286

6. Using this [online post](https://wilkelab.org/classes/SDS348/2016_spring/projects/project1/project1_hints.html) proceed to divide the data set into three evenly-sized groups corresponding to *small*, *medium*, and *large* sized firms in term of ```lsales```, and add dummy variables ```small```, ```medium```, and ```large``` to your data set.

In [8]:
# divide data into three evenly-sized groups corresponding to small, medium and large sized firms (terms of lsales)
ceo_salary <- ceo_salary %>% mutate(three_groups = ntile(lsales,3))

# add dummy variables small, medium, large to data set
# create dummy variables
ceo_salary$small <- rep(0,nrow(ceo_salary))
ceo_salary$medium <- rep(0, nrow(ceo_salary))
ceo_salary$large <- rep(0, nrow(ceo_salary))

# create for loop for observations
for(i in 1:nrow(ceo_salary)) {
    # create if else statements to put 1's in
    if(ceo_salary$three_groups[i] == 1)
        ceo_salary$small[i] <- 1
    
    else if (ceo_salary$three_groups[i] == 2)
        ceo_salary$medium[i] <- 1
        
    else if (ceo_salary$three_groups[i] == 3)
        ceo_salary$large[i] <- 1
}
  
# remove three_groups column
ceo_salary <- subset(ceo_salary, select = -(three_groups))

# show dataset
head(ceo_salary)

salary age college grad comten ceoten sales profits mktval lsalary  lsales  
1 1161   49  1       1     9      2      6200 966     23200  7.057037 8.732305
2  600   43  1       1    10     10       283  48      1100  6.396930 5.645447
3  379   51  1       1     9      3       169  40      1100  5.937536 5.129899
4  651   55  1       0    22     22      1100 -54      1000  6.478509 7.003066
5  497   44  1       1     8      6       351  28       387  6.208590 5.860786
6 1067   64  1       1     7      7     19000 614      3900  6.972606 9.852194
  lmktval   comtensq ceotensq profmarg  d_comten_sq  d_ceoten_sq small medium
1 10.051908  81        4      15.580646 173.56530557  33.9373986 0     0     
2  7.003066 100      100      16.961130 148.21646836   4.7280963 1     0     
3  7.003066  81        9      23.668638 173.56530557  23.2862358 1     0     
4  6.907755 484      484      -4.909091   0.03042185 200.9141428 0     1     
5  5.958425  64       36       7.977208 200.91414278   3.3327474 1     0     
6  8.268732  49       49       3.231579 230.26297999   0.6815846 0     0     
  large
1 1    
2 0    
3 0    
4 0    
5 0    
6 1

#### <font color='red'>Comments:</font> 

Again, it is a good practice to try to avoid loops. In this case I use the ```model.matrix``` function along with the ```factor``` option to avoid the loop. 

7. Estimate $E[\texttt{salary}|\texttt{small}=1]$, $E[\texttt{salary}|\texttt{medium}=1]$, and $E[\texttt{salary}|\texttt{large}=1]$, and print these values.

In [9]:
# estimate E[salary|small=1]
round(mean((filter(ceo_salary, small==1))$salary),3)

# estimate E[salary|medium=1]
round(mean((filter(ceo_salary, medium==1))$salary),3)

# estimate E[salary|large=1]
round(mean((filter(ceo_salary, large==1))$salary),3)


[1] 617.237

[1] 836.441

[1] 1143.915

8. Now consider the following model

$$
\texttt{lsalary}=\beta_0+\beta_1\texttt{grad}+\beta_2\texttt{medium}+\beta_3\texttt{large}+\beta_4\texttt{lmktval}+\beta_5\texttt{comten}+\beta_6\texttt{ceoten}+\beta_7\texttt{d}\_\texttt{comten}\_\texttt{sq}+\beta_8\texttt{d}\_\texttt{ceoten}\_\texttt{sq}+e\text{.}
$$

Is there perfect multicollinearity?

In [18]:
# add log salary to ceo_salary dataset
ceo_salary$lsalary <- log(ceo_salary$salary)

# specify the model
model <- as.formula(paste("lsalary", 
            paste(c("grad","medium","large","lmktval","comten","ceoten","d_comten_sq","d_ceoten_sq"), collapse= " + "),
            sep = " ~ " ))

# create the design matrix
matrix <- model.matrix(model, data=ceo_salary)

# calculate & print E[xx']
X <- t(matrix) %*% matrix/nrow(matrix)

# get rank and dimensions of E[xx']
dimension <- dim(X)
rank <- qr(X)$rank

# check if the dimensions of X match rank
if(rank == dimension) {
    print("There is no perfect multicollinearity")
    } else {
    print("There is perfect collinearity") }

Warning message in if (rank == dimension) {:
"the condition has length > 1 and only the first element will be used"

[1] "There is no perfect multicollinearity"


#### <font color='red'>Comments:</font> 

All here is correct. One small comment: note that ```length(dimension)``` is two and ```length(rank)```. The warning message is saying that it only the first element will be used. In this case it does not matter what dimension to use because the matrix $\mathbf{X}^{\prime}\mathbf{X}$ is symmetric but it may be important in other situations. Because the minimum dimension is used to calculate whether a matrix is complete rank, a good practice would be to use ```min(dimension[1],dimension[2])``` in the ```if``` statement.

9. You are told that a *better* model would be one where *each* dummy variable in the previous specification is interacted with the remaining features. Print the dimensions of the corresponding **design matrix**, $\mathbf{X}$, for this case and check that $\mathbf{X}^{\prime}\mathbf{X}$ is invertible for this model.

In [20]:
# create character vectors of terms needed
dummies <- c("medium", "large", "grad")
predictors <- c("lmktval","comten","ceoten","d_comten_sq","d_ceoten_sq")

# create new formula
f <- as.formula(paste("lsalary", paste(outer(dummies, predictors, FUN= "paste", sep=":"), collapse = " + "), sep=" ~ "))

# create matrix
interact_matrix <- model.matrix(f, data=ceo_salary)

# check X'X is invertible
# regular matrix
t(interact_matrix) %*% interact_matrix/nrow(interact_matrix)

# invertible matrix
solve(t(interact_matrix) %*% interact_matrix/nrow(interact_matrix))


(Intercept) medium:lmktval lmktval:large lmktval:grad
(Intercept)         1.000000     2.353654       2.854037      4.014300  
medium:lmktval      2.353654    16.764356       0.000000      7.033834  
lmktval:large       2.854037     0.000000      24.754021     16.612440  
lmktval:grad        4.014300     7.033834      16.612440     31.134387  
medium:comten       8.073446    56.760826       0.000000     21.399392  
large:comten        8.163842     0.000000      70.179258     44.519386  
grad:comten        10.553672    21.399392      44.519386     81.592298  
medium:ceoten       2.881356    20.212002       0.000000      6.941588  
large:ceoten        2.446328     0.000000      21.101470     14.312509  
grad:ceoten         3.858757     6.941588      14.312509     29.141589  
medium:d_comten_sq 39.859306   281.770126       0.000000     90.679245  
large:d_comten_sq  47.307999     0.000000     406.166934    265.457246  
grad:d_comten_sq   72.173034    90.679245     265.457246    540.562349  
medium:d_ceoten_sq 24.220625   167.098058       0.000000     39.884662  
large:d_ceoten_sq  13.893730     0.000000     122.902544     77.115124  
grad:d_ceoten_sq   20.010112    39.884662      77.115124    153.890594  
                   medium:comten large:comten grad:comten medium:ceoten
(Intercept)           8.073446      8.163842    10.55367    2.881356   
medium:lmktval       56.760826      0.000000    21.39939   20.212002   
lmktval:large         0.000000     70.179258    44.51939    0.000000   
lmktval:grad         21.399392     44.519386    81.59230    6.941588   
medium:comten       234.005650      0.000000    79.94915   77.084746   
large:comten          0.000000    245.463277   148.94915    0.000000   
grad:comten          79.949153    148.949153   279.08475   23.649718   
medium:ceoten        77.084746      0.000000    23.64972   48.903955   
large:ceoten          0.000000     67.011299    42.12429    0.000000   
grad:ceoten          23.649718     42.124294    86.58757   12.937853   
medium:d_comten_sq 1075.845925      0.000000   251.29002  398.835784   
large:d_comten_sq     0.000000   1062.150080   609.74255    0.000000   
grad:d_comten_sq    251.290017    609.742551  1126.15131   71.814853   
medium:d_ceoten_sq  795.396049      0.000000   141.63504  615.524261   
large:d_ceoten_sq     0.000000    408.609941   236.65334    0.000000   
grad:d_ceoten_sq    141.635044    236.653338   476.53155   72.268407   
                   large:ceoten grad:ceoten medium:d_comten_sq
(Intercept)          2.446328     3.858757    39.85931        
medium:lmktval       0.000000     6.941588   281.77013        
lmktval:large       21.101470    14.312509     0.00000        
lmktval:grad        14.312509    29.141589    90.67924        
medium:comten        0.000000    23.649718  1075.84592        
large:comten        67.011299    42.124294     0.00000        
grad:comten         42.124294    86.587571   251.29002        
medium:ceoten        0.000000    12.937853   398.83578        
large:ceoten        31.768362    20.802260     0.00000        
grad:ceoten         20.802260    47.881356    71.81485        
medium:d_comten_sq   0.000000    71.814853  8938.53221        
large:d_comten_sq  341.173095   189.767702     0.00000        
grad:d_comten_sq   189.767702   406.068514  2488.00654        
medium:d_ceoten_sq   0.000000    72.268407  4994.18142        
large:d_ceoten_sq  207.385945   129.132305     0.00000        
grad:d_ceoten_sq   129.132305   270.735181   500.82293        
                   large:d_comten_sq grad:d_comten_sq medium:d_ceoten_sq
(Intercept)           47.3080           72.17303         24.22063       
medium:lmktval         0.0000           90.67924        167.09806       
lmktval:large        406.1669          265.45725          0.00000       
lmktval:grad         265.4572          540.56235         39.88466       
medium:comten          0.0000          251.29002        795.39605       
large:comten        1062.1501          609.74

(Intercept)  medium:lmktval lmktval:large lmktval:grad
(Intercept)         4.094726398 -0.380202945   -0.211252299  -0.223046910
medium:lmktval     -0.380202945  0.621387356    0.151057984  -0.206307393
lmktval:large      -0.211252299  0.151057984    0.514564088  -0.285885071
lmktval:grad       -0.223046910 -0.206307393   -0.285885071   0.541721149
medium:comten      -0.041110304 -0.092277143   -0.018830208   0.037317001
large:comten       -0.061329613 -0.014111816   -0.097271315   0.060267877
grad:comten         0.050736377  0.025647068    0.051215793  -0.095006424
medium:ceoten       0.025024880 -0.166264335   -0.031791359   0.061191936
large:ceoten        0.075582561 -0.039246407   -0.073281512   0.024319765
grad:ceoten        -0.120574764  0.082414910    0.049412100  -0.083991613
medium:d_comten_sq  0.001151399 -0.004034064   -0.001018034   0.001833650
large:d_comten_sq   0.001801023 -0.001279868   -0.006397592   0.003600393
grad:d_comten_sq   -0.005072908  0.003127186    0.004286872  -0.007797259
medium:d_ceoten_sq -0.001249124  0.008447704    0.001691304  -0.003116940
large:d_ceoten_sq  -0.003271602  0.002070207    0.005893618  -0.002204217
grad:d_ceoten_sq    0.005725806 -0.003818704   -0.003419846   0.003650154
                   medium:comten large:comten  grad:comten   medium:ceoten
(Intercept)        -4.111030e-02 -6.132961e-02  5.073638e-02  2.502488e-02
medium:lmktval     -9.227714e-02 -1.411182e-02  2.564707e-02 -1.662643e-01
lmktval:large      -1.883021e-02 -9.727132e-02  5.121579e-02 -3.179136e-02
lmktval:grad        3.731700e-02  6.026788e-02 -9.500642e-02  6.119194e-02
medium:comten       3.325497e-02  7.658727e-03 -1.158903e-02  2.665429e-03
large:comten        7.658727e-03  3.793683e-02 -2.059646e-02 -1.843347e-03
grad:comten        -1.158903e-02 -2.059646e-02  3.249523e-02  1.454230e-03
medium:ceoten       2.665429e-03 -1.843347e-03  1.454230e-03  1.675035e-01
large:ceoten       -5.641017e-04 -1.715514e-02  9.489883e-03  4.088812e-02
grad:ceoten        -9.787850e-04  8.100670e-03 -9.644042e-03 -6.974593e-02
medium:d_comten_sq -6.762061e-04 -8.284892e-05  8.891578e-05  1.065015e-03
large:d_comten_sq  -1.656230e-05  6.099753e-04 -4.029479e-04 -2.571991e-05
grad:d_comten_sq   -9.971302e-05 -4.394115e-04  9.542546e-04 -4.345818e-04
medium:d_ceoten_sq -6.839580e-04 -7.322740e-05  1.798088e-04 -6.915146e-03
large:d_ceoten_sq  -1.240538e-04 -7.037140e-04  4.606319e-04 -1.512109e-03
grad:d_ceoten_sq    2.118337e-04  4.888636e-04 -5.708075e-04  2.732399e-03
                   large:ceoten  grad:ceoten   medium:d_comten_sq
(Intercept)         7.558256e-02 -0.1205747639  1.151399e-03     
medium:lmktval     -3.924641e-02  0.0824149098 -4.034064e-03     
lmktval:large      -7.328151e-02  0.0494121005 -1.018034e-03     
lmktval:grad        2.431977e-02 -0.0839916131  1.833650e-03     
medium:comten      -5.641017e-04 -0.0009787850 -6.762061e-04     
large:comten       -1.715514e-02  0.0081006701 -8.284892e-05     
grad:comten         9.489883e-03 -0.0096440424  8.891578e-05     
medium:ceoten       4.088812e-02 -0.0697459320  1.065015e-03     
large:ceoten        1.999418e-01 -0.0897074304 -5.141330e-05     
grad:ceoten        -8.970743e-02  0.1501049654 -2.186917e-04     
medium:d_comten_sq -5.141330e-05 -0.0002186917  3.244752e-04     
large:d_comten_sq   3.092659e-04 -0.0002005113  5.182002e-05     
grad:d_comten_sq   -3.214892e-05  0.0009232463 -8.693775e-05     
medium:d_ceoten_sq -1.607567e-03  0.0028003583 -9.476224e-05     
large:d_ceoten_sq  -1.222330e-02  0.0054389309 -1.594038e-05     
grad:d_ceoten_sq    5.521673e-03 -0.0092120661  1.962541e-05     
                   large:d_comten_sq grad:d_comten_sq medium:d_ceoten_sq
(Intercept)         1.801023e-03     -5.072908e-03    -1.249124e-03     
medium:lmktval     -1.279868e-03      3.127186e-03     8.447704e-03     
lmktval:large      -6.397592e-03      4.286872e-03     1.691304e-03     
lmktval:grad        3.600393e-03     -7.797259e-03    -3.116940e-03     

#### <font color='red'>Comments:</font> 

In this exercise you have to consider the base regression from question 8 and add the interaction between dummies and the remaining features. See Wooldridge chapter 7 for a good explanation of why you need both the base regression and the interactions in your design matrix. You did not print the dimensions of the design matrix, with all the required variables the dimension should be $117 \times 24$.

10. Estimate the model in (9) above by OLS and report standard errors that are robust to the presence of heteroskedasticity of unknown form.

In [51]:
# estimate model
(reg <- lm(f, data=ceo_salary))

# install and load sandwich package
if(!require(sandwich)) install.packages("sandwich")
library(sandwich)

# report robust standard errors
data.frame("HC1" = round(sqrt(diag(vcovHC(reg, type="HC1"))),6))


Call:
lm(formula = f, data = ceo_salary)

Coefficients:
       (Intercept)      medium:lmktval       lmktval:large        lmktval:grad  
         6.2470238           0.0291468           0.0923940           0.0136383  
     medium:comten        large:comten         grad:comten       medium:ceoten  
        -0.0067311          -0.0063486          -0.0046799           0.0391360  
      large:ceoten         grad:ceoten  medium:d_comten_sq   large:d_comten_sq  
         0.0348888          -0.0059647           0.0006429           0.0007033  
  grad:d_comten_sq  medium:d_ceoten_sq   large:d_ceoten_sq    grad:d_ceoten_sq  
        -0.0004954          -0.0013247          -0.0062971           0.0012830  


HC1     
(Intercept)        0.086771
medium:lmktval     0.040208
lmktval:large      0.024272
lmktval:grad       0.025321
medium:comten      0.006918
large:comten       0.006892
grad:comten        0.006345
medium:ceoten      0.018686
large:ceoten       0.015221
grad:ceoten        0.014564
medium:d_comten_sq 0.000986
large:d_comten_sq  0.000704
grad:d_comten_sq   0.000677
medium:d_ceoten_sq 0.001008
large:d_ceoten_sq  0.001742
grad:d_ceoten_sq   0.001417

#### <font color='red'>Comments:</font> 

Again, in this exercise you needed to include the 24 required explanatory variables in the regression.